controllers/volumereplicationgroup_controller.go

```diff
@@ -409,94 +421,170 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
+
+	// if the PVC is not bound yet, don't proceed
+	if pvc.Status.Phase != corev1.ClaimBound {
+		log.Info("Requeuing as PersistentVolumeClaim is not bound", "pvcPhase", pvc.Status.Phase)
 
 		return requeue
 	}
 
-	if requeueResult := v.handlePersistentVolumes(); requeueResult {
-		v.log.Info("Requeuing resource")
+	// TODO: Should we ignore PVCs marked for deletion?
+
+	// Check if PVC is already protected
+	pvcProtected, err := v.ensurePVCVRProtected(pvc, log)
+	if err != nil {
+		log.Info("Requeuing as determining PersistentVolumeClaim protection failed", "errorValue", err)
+
+		return requeue
+	}
+
+	if pvcProtected {
+		return !requeue
+	}
+
+	// Add `vr-protecting` finalizer to PVC
+	if err = v.addProtectingFinalizerToPVC(pvc, log); err != nil {
+		log.Info("Requeuing as protecting PersistentVolumeClaim failed", "errorValue", err)
+
+		return requeue
+	}
+
+	// Change PV `reclaimPolicy` to "Retain"
+	if err = v.retainPVForPVC(*pvc, log); err != nil {
+		log.Info("Requeuing as retaining PersistentVolume failed", "errorValue", err)
 
 		return requeue
 	}
 
 	return !requeue
 }
 
-func (v *VRGInstance) createVolumeReplicationResources() bool {
-	requeue := false
+// retainPVForPVC updates the PV reclaim policy to retain for a given PVC
+func (v *VRGInstance) retainPVForPVC(pvc corev1.PersistentVolumeClaim, log logr.Logger) error {
+	// Get PV bound to PVC
+	pv := &corev1.PersistentVolume{}
+	pvObjectKey := client.ObjectKey{
+		Name: pvc.Spec.VolumeName,
+	}
 
-	for _, pvc := range v.pvcList.Items {
-		pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
-		log := v.log.WithValues("pvc", pvcNamespacedName.String())
+	if err := v.reconciler.Get(v.ctx, pvObjectKey, pv); err != nil {
+		log.Error(err, "Failed to get PersistentVolume", "volumeName", pvc.Spec.VolumeName)
 
-		// if the PVC is not bound yet, dont proceed.
-		if pvc.Status.Phase != corev1.ClaimBound {
-			log.Info("Requeuing as PersistentVolumeClaim is not bound",
-				"pvcPhase", pvc.Status.Phase)
+		return fmt.Errorf("failed to get PersistentVolume resource (%s) for"+
+			" PersistentVolumeClaim resource (%s/%s) belonging to VolumeReplicationGroup (%s/%s), %w",
+			pvc.Spec.VolumeName, pvc.Name, pvc.Namespace, v.instance.Name, v.instance.Namespace, err)
+	}
 
-			requeue = true
+	// Check reclaimPolicy of PV, if already set to retain
+	if pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimRetain {
+		return nil
+	}
 
-			// continue processing other PVCs and return the need to requeue
-			continue
-		}
+	// if not retained, retain PV, and add an annotation to denote this is updated for VR needs
+	pv.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+	pv.ObjectMeta.Annotations[pvVRAnnotationRetentionKey] = pvVRAnnotationRetentionValue
 
-		volRep := &volrep.VolumeReplication{}
+	if err := v.reconciler.Update(v.ctx, pv); err != nil {
+		log.Error(err, "Failed to update PersistentVolume reclaim policy")
```

Comment:
(nit) is it better to also print the name of the persistent volume and the pvc that it belongs to?

Reply:
I see that we are returning that information below. So, the above comment can be ignored.